### PR TITLE
Update `ContrastiveTensionLoss` and `ContrastiveTensionLossInBatchNegatives`

### PR DESCRIPTION
**Pull request by omkar-334**
#3627

See [comment](https://github.com/huggingface/sentence-transformers/pull/3627/#issuecomment-3817696102) 

> If you'd like, then some changes can be made to ContrastiveTensionLossInBatchNegatives to make it a bit stronger:
>  1. I feel like it should not be documented with model.fit, no longer mention the ContrastiveTensionDataLoader,
>  2. it should accept single sentences with none labels. It should then use sentence_features[0] for self.model1 and self.model2, it's a bit weird that it asks for the same text twice.

Created a new PR for this since the scope felt a bit too large for a follow-up commit

### CODE REVIEW

File: sentence_transformers/losses/ContrastiveTensionLoss.py
Comment (omkar-334) [inline sentence_transformers/losses/ContrastiveTensionLoss.py:122]: Should we also edit the docstring for ContrastiveTensionLoss? I think this will need the PR for `ContrastiveTensionDataLoader`-alternative to be merged first

File: sentence_transformers/losses/ContrastiveTensionLoss.py [hunk 1]
```diff
@@ -151,20 +150,34 @@ def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_f
...
-                    InputExample(texts=['This is the 2nd sentence', 'This is the 2nd sentence']),
-                    InputExample(texts=['This is the 3rd sentence', 'This is the 3rd sentence']),
-                    InputExample(texts=['This is the 4th sentence', 'This is the 4th sentence']),
-                    InputExample(texts=['This is the 5th sentence', 'This is the 5th sentence']),
+                    InputExample(texts=['This is the 1st sentence']),
+                    InputExample(texts=['This is the 2nd sentence']),
+                    InputExample(texts=['This is the 3rd sentence']),
+                    InputExample(texts=['This is the 4th sentence']),
+                    InputExample(texts=['This is the 5th sentence']),
                 ]
```
Comment (omkar-334) [inline sentence_transformers/losses/ContrastiveTensionLoss.py]: Why do we have `train_examples` twice here? Should i remove the first variable?

File: sentence_transformers/losses/ContrastiveTensionLoss.py [hunk 2]
```diff
@@ -174,9 +187,9 @@ def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_f
         self.logit_scale = nn.Parameter(torch.ones([]) * np.log(scale))
 
     def forward(self, sentence_features: Iterable[dict[str, Tensor]], labels: Tensor) -> Tensor:
-        sentence_features1, sentence_features2 = tuple(sentence_features)
-        embeddings_a = self.model1(sentence_features1)["sentence_embedding"]  # (bsz, hdim)
-        embeddings_b = self.model2(sentence_features2)["sentence_embedding"]
+        sentence_features = sentence_features[0]
+        embeddings_a = self.model1(sentence_features)["sentence_embedding"]  # (bsz, hdim)
+        embeddings_b = self.model2(sentence_features)["sentence_embedding"]
```
Comment (tomaarsen) [inline sentence_transformers/losses/ContrastiveTensionLoss.py:258]: Perfect!

File: sentence_transformers/losses/ContrastiveTensionLoss.py (hunk 1 above)
Comment (tomaarsen) [reply sentence_transformers/losses/ContrastiveTensionLoss.py]: Yeah I think the first one was always wrong (because you were meant to provide two identical texts).

And I think the cleanest is to immediately make a Dataset, e.g. like:
```python
                train_dataset = Dataset.from_dict({
                    "sentence": [
                       "It's nice weather outside today.",
                       "He drove to work.",
                       "It's so sunny.",
                    ]
                })
```
Then we don't need the InputExample (which is also a bit dated and I'd love to slowly stop using)

File: sentence_transformers/losses/ContrastiveTensionLoss.py [hunk 3]
```diff
@@ -114,8 +114,7 @@ def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_f
         are sampled from the batch. Using in-batch negative sampling gives a stronger training signal than the original :class:`ContrastiveTensionLoss`.
         The performance usually increases with increasing batch sizes.
 
-        Note that you should not use the `ContrastiveTensionDataLoader` for this loss, but just a normal DataLoader with `InputExample` instances.
-        The two texts of each `InputExample` instance should be identical.
+        Each `InputExample` instance must contain a single text.
```
Comment (tomaarsen) [inline sentence_transformers/losses/ContrastiveTensionLoss.py]: I think we can mention instead e.g. a Dataset with one text column:

```suggestion
        The training :class:`datasets.Dataset` must contain one text column.
```
P.s. the 
```
:class:`datasets.Dataset`
```
is from Sphinx autodoc and will automatically link to the Datasets documentation.

File: sentence_transformers/losses/ContrastiveTensionLoss.py
Comment (tomaarsen) [reply sentence_transformers/losses/ContrastiveTensionLoss.py:122]: I think it would be nice to upgrade that class too, like mentioned in https://github.com/huggingface/sentence-transformers/pull/3625#issuecomment-3817649758, but that can be implemented in this PR, the PR where I posted the comment, or a new PR, in case you want to work on that as well. Whatever you prefer 🤗

File: sentence_transformers/losses/ContrastiveTensionLoss.py [hunk 4]
```diff
@@ -49,28 +49,39 @@ class ContrastiveTensionLoss(nn.Module):
...
+                    "It's so sunny.",
+                ]
+                "sentence_B": [
+                    "It's nice weather outside today.",
+                    "He drove to work.",
+                    "It's so sunny.",
+                ]
+                })
+
+            train_dataloader = ContrastiveTensionDataLoader(train_dataset, batch_size=3, pos_neg_ratio=3)
```
Comment (omkar-334) [inline sentence_transformers/losses/ContrastiveTensionLoss.py]: The only thing left is this... Should we update this with https://github.com/huggingface/sentence-transformers/pull/3625#issuecomment-3828489460
or let this be?

Comment (omkar-334) [reply sentence_transformers/losses/ContrastiveTensionLoss.py]: except this, I think this PR along with #3264 and #3638 are ready to be merged!

Comment (tomaarsen) [reply sentence_transformers/losses/ContrastiveTensionLoss.py]: Yes! I've updated it in 458cedd8d8b55e5fce3fd164f6d0ca0fde6ff806, alongside some other changes that I noticed. For example, the `label` approach required identical padding length across  the two inputs, but that wasn't always the case.

Comment (omkar-334) [reply sentence_transformers/losses/ContrastiveTensionLoss.py]: hmm, got it. It does look better now.

**Comment by omkar-334** [comment]
> I think this is ready now! What do you think?

Yes!
